### PR TITLE
add functionality and checks to mapping file validation

### LINKED ISSUE #41: Add additional checks to validation script
Already implemented in develop branch
- [x] check for empty code name
- [x] check for code not in respective dataset file

Next steps:
- [x] Read dataset names directly from reference Synapse table for now to match script functionality 
- [x] Allow input to be Synapse ID where the mapping file is stored or a path to the local mapping file
- [x] Update readme on validation script usage
- [x] Confirm 'datasets' match map to derived variables
- [x] Verify cohort specific release status with scope of release

### PR DESCRIPTION
**Pull request by hhunterzinck**
Fixes #41 

Added:
- synapse id or file path input
- check against datasets on Synapse
- check for ambiguous release status in mapping file
- check release status again scope of release

### CODE REVIEW

File: scripts/validate_map.py [hunk 1]
```diff
@@ -18,34 +20,70 @@
 import yaml
 
 
-def check_code_name_empty(df: pd.DataFrame, syn: Synapse, config: dict) -> list:
+def get_codes_to_remove(codes: list) -> list:
```
Comment (thomasyu888) [inline scripts/validate_map.py]: Use typing.List

File: scripts/validate_map.py [hunk 2]
```diff
@@ -18,34 +20,70 @@
...
+    for code in codes:
+        if bool(re.match(r"^.+[*]$", str(code).strip())):
+            code_remove.append(code)
+        elif pd.isna(code):
+            code_remove.append(code)
+    return code_remove
+
+
+def check_code_name_empty(
+    df: pd.DataFrame, syn: Synapse, config: dict, cohort: str, release: str
```
Comment (thomasyu888) [inline scripts/validate_map.py]: Use typing.Dict

File: scripts/validate_map.py [hunk 3]
```diff
@@ -18,34 +20,70 @@
...
     absent = []
-    for dataset in config["dataset"]:
-        data = pd.read_csv(
-            syn.get(config["dataset"][dataset]["id"])["path"], low_memory=False
-        )
+
+    query = f'SELECT id, dataset FROM {config["synapse"]["dataset"]["id"]} WHERE dataset IS NOT NULL'
+    res = syn.tableQuery(query)
+
+    for row in res:
```
Comment (thomasyu888) [inline scripts/validate_map.py:81]: Does this work? I this it's missing res.asRowSet()?

**Review by thomasyu888 — CHANGES_REQUESTED** [review]

File: scripts/validate_map.py (hunk 2 above)
Comment (hhunterzinck) [reply scripts/validate_map.py]: Replaced dict with Dict for both arguments and return hints

File: scripts/validate_map.py (hunk 1 above)
Comment (hhunterzinck) [reply scripts/validate_map.py]: Replaced list with List for both arguments and return hints

File: scripts/validate_map.py (hunk 3 above)
Comment (hhunterzinck) [reply scripts/validate_map.py:81]: For each iteration, `row` is a list with two elements, which is what I was aiming for.  Is this not the correct way to iterate over the results of Synapse tableQuery?

Comment (thomasyu888) [reply scripts/validate_map.py:81]: I didn't realize tableQuery could utilize the built in iteration function. I thought you had to specify .asDataFrame or .asRowSet before you could touch the values.

File: scripts/validate_map.py [hunk 4]
```diff
@@ -62,27 +100,103 @@ def check_code_name_absent(df: pd.DataFrame, syn: Synapse, config: dict) -> list
...
+    """
+    map_status = df[cohort].str.lower()
+    map_type = df["data_type"].str.lower()
+    map_codes = df.loc[
+        ((map_status == "y") & ((map_type == "derived") | (map_type == "curated")))
+    ]["code"]
+
+    column_name = config["column_name"]["sor_cbio"][cohort][release]
+    file_sor = syn.get(config["synapse"]["sor"]["id"])["path"]
+    sor = pd.read_excel(file_sor, engine="openpyxl", sheet_name=1)
```
Comment (thomasyu888) [inline scripts/validate_map.py:178]: Why openpyxl?

File: scripts/config.yaml [hunk 5]
```diff
@@ -1,40 +1,31 @@
...
+    id: syn22296821
+    name: Data files for derived variables
+    type: table
+    description: Links derived variable Synapse entities to dataset names.
+  sor: 
+    id: syn22294851
+    name: GENIE_BPC_Release_Scope.xlsx
+    type: file
+    description: stats team Scope of Release for GENIE BPC
+column_name:
```
Comment (thomasyu888) [inline scripts/config.yaml]: These are the column names in the excel file?

File: scripts/validate_map.py (hunk 4 above)
Comment (hhunterzinck) [reply scripts/validate_map.py:178]: Default engine for pandas.read_excel is xlrd and doesn't support reading xlsx, which is what file_sor is.  

https://stackoverflow.com/questions/65254535/xlrd-biffh-xlrderror-excel-xlsx-file-not-supported

File: scripts/config.yaml (hunk 5 above)
Comment (hhunterzinck) [reply scripts/config.yaml]: yes.  discussion with file providers indicates that naming hasn't been and probably won't be very systematic, so I just spelled each out explicitly for the cohort-release pairs.

I realized I had one column name specified incorrectly, though, so I just updated it.

Comment (thomasyu888) [reply scripts/config.yaml]: Should this section just be under the synapse id configuration itself?

Seems like this is metadata about the file. What if we needed column names for other files?

Comment (hhunterzinck) [reply scripts/config.yaml]: Great idea.  I've moved the column name specifications under the file and updated the configuration keys respectively in the main script.

**Review by thomasyu888 — APPROVED** [review]

**Pull request merged**